File: add-ons/redis-cache/api-fastapi/cache.py
"""
Cache Utility Module

Provides caching functionality with Redis (optional) or in-memory fallback.
Install Redis support: pip install redis>=5.0.0

Usage:
    from app.core.cache import cache, cached

    # Set a value (TTL in seconds)
    await cache.set("key:123", value, ttl=600)

    # Get a value
    value = await cache.get("key:123")

    # Delete a value
    await cache.delete("key:123")

    # Delete all keys matching pattern (supports * wildcard)
    await cache.delete_pattern("user:*")        # All user keys
    await cache.delete_pattern("product:*")     # All product keys

    # Use decorator for function caching
    @cached(ttl=300)
    async def get_data(data_id: str):
        # This result will be cached for 5 minutes
        return await fetch_from_database(data_id)
"""

from __future__ import annotations

import json
import hashlib
import functools
from typing import Any, Callable, TypeVar, ParamSpec
from datetime import datetime, timedelta

# Try to import Redis, fall back to in-memory if not available
try:
    import redis.asyncio as redis
    REDIS_AVAILABLE = True
except ImportError:
    REDIS_AVAILABLE = False

from app.core.config import settings


P = ParamSpec("P")
T = TypeVar("T")


class InMemoryCache:
    """Simple in-memory cache for development/testing."""

    def __init__(self) -> None:
        self._store: dict[str, tuple[Any, datetime | None]] = {}

    async def get(self, key: str) -> Any | None:
        """Get value from cache."""
        if key not in self._store:
            return None

        value, expires_at = self._store[key]

        if expires_at and datetime.now() > expires_at:
            del self._store[key]
            return None

        return value

    async def set(self, key: str, value: Any, ttl: int | None = None) -> None:
        """Set value in cache with optional TTL (seconds)."""
        expires_at = datetime.now() + timedelta(seconds=ttl) if ttl else None
        self._store[key] = (value, expires_at)

    async def delete(self, key: str) -> None:
        """Delete value from cache."""
        self._store.pop(key, None)

    async def delete_pattern(self, pattern: str) -> None:
        """Delete all keys matching pattern (supports * wildcard)."""
        # Simple wildcard support for in-memory cache
        if "*" in pattern:
            prefix, suffix = pattern.split("*", 1)
            keys_to_delete = [
                key for key in list(self._store.keys())
                if key.startswith(prefix) and (not suffix or key.endswith(suffix))
            ]
            for key in keys_to_delete:
                self._store.pop(key, None)
        else:
            # Exact match
            self._store.pop(pattern, None)

    async def exists(self, key: str) -> bool:
        """Check if key exists and is not expired."""
        return await self.get(key) is not None

    async def clear(self) -> None:
        """Clear all cached values."""
        self._store.clear()

    async def close(self) -> None:
        """Close connection (no-op for in-memory)."""
        pass


class RedisCache:
    """Redis-backed cache for production."""

    def __init__(self, url: str) -> None:
        self._url = url
        self._client: redis.Redis | None = None

    async def _get_client(self) -> redis.Redis:
        """Lazy initialization of Redis client."""
        if self._client is None:
            self._client = redis.from_url(
                self._url,
                encoding="utf-8",
                decode_responses=True,
            )
        return self._client

    async def get(self, key: str) -> Any | None:
        """Get value from cache."""
        client = await self._get_client()
        value = await client.get(key)

        if value is None:
            return None

        try:
            return json.loads(value)
        except json.JSONDecodeError:
            return value

    async def set(self, key: str, value: Any, ttl: int | None = None) -> None:
        """Set value in cache with optional TTL (seconds)."""
        client = await self._get_client()
        serialized = json.dumps(value) if not isinstance(value, str) else value

        if ttl:
            await client.setex(key, ttl, serialized)
        else:
            await client.set(key, serialized)

    async def delete(self, key: str) -> None:
        """Delete value from cache."""
        client = await self._get_client()
        await client.delete(key)

    async def delete_pattern(self, pattern: str) -> None:
        """Delete all keys matching pattern (supports * wildcard via SCAN)."""
        client = await self._get_client()
        deleted_count = 0
        
        async for key in client.scan_iter(match=pattern):
            await client.delete(key)
            deleted_count += 1
        
        if deleted_count > 0:
            import logging
            logger = logging.getLogger(__name__)
            logger.debug(f"Deleted {deleted_count} cache keys matching pattern: {pattern}")

    async def exists(self, key: str) -> bool:
        """Check if key exists."""
        client = await self._get_client()
        return await client.exists(key) > 0

    async def clear(self) -> None:
        """Clear all cached values (use with caution)."""
        client = await self._get_client()
        await client.flushdb()

    async def close(self) -> None:
        """Close Redis connection."""
        if self._client:
            await self._client.close()


def _create_cache() -> InMemoryCache | RedisCache:
    """Create appropriate cache backend based on configuration."""
    redis_url = getattr(settings, "REDIS_URL", None)

    if redis_url and REDIS_AVAILABLE:
        return RedisCache(redis_url)

    if redis_url and not REDIS_AVAILABLE:
        import warnings
        warnings.warn(
            "REDIS_URL is set but redis package is not installed. "
            "Using in-memory cache. Install with: pip install 'redis>=5.0.0'"
        )

    return InMemoryCache()


# Global cache instance
cache = _create_cache()


def cached(
    ttl: int = 300,
    key_prefix: str = "",
    key_builder: Callable[..., str] | None = None,
) -> Callable[[Callable[P, T]], Callable[P, T]]:
    """
    Decorator to cache function results.
    
    Args:
        ttl: Time to live in seconds (default: 5 minutes)
        key_prefix: Prefix for cache key
        key_builder: Custom function to build cache key from args
    
    Usage:
        @cached(ttl=600)
        async def get_data(data_id: str) -> Data:
            return await db.get_data(data_id)
        
        @cached(ttl=60, key_prefix="items")
        async def list_items(filters: dict, limit: int) -> list[Item]:
            return await db.list_items(filters, limit)
    """

    def decorator(func: Callable[P, T]) -> Callable[P, T]:
        @functools.wraps(func)
        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            # Build cache key
            if key_builder:
                cache_key = key_builder(*args, **kwargs)
            else:
                # Default key: prefix:function_name:hash(args)
                arg_hash = hashlib.md5(
                    json.dumps((args, kwargs), sort_keys=True, default=str).encode()
                ).hexdigest()[:16]
                prefix = key_prefix or func.__module__
                cache_key = f"{prefix}:{func.__name__}:{arg_hash}"

            # Try to get from cache
            cached_value = await cache.get(cache_key)
            if cached_value is not None:
                return cached_value

            # Call function and cache result
            result = await func(*args, **kwargs)
            await cache.set(cache_key, result, ttl=ttl)
            return result

        return wrapper

    return decorator


def invalidate_cache(pattern: str) -> Callable[[Callable[P, T]], Callable[P, T]]:
    """
    Decorator to invalidate cache after function execution.
    
    Usage:
        @invalidate_cache("item:*")
        async def update_item(item_id: str, data: dict) -> Item:
            # After this runs, cache entries matching "item:*" should be cleared
            return await db.update_item(item_id, data)
    """

    def decorator(func: Callable[P, T]) -> Callable[P, T]:
        @functools.wraps(func)
        async def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
            result = await func(*args, **kwargs)

            # Delete all keys matching pattern (supports wildcards)
            await cache.delete_pattern(pattern)

            return result

        return wrapper

    return decorator

File: add-ons/redis-cache/api-express/cache.ts
/**
 * Cache Utility Module
 *
 * Provides caching functionality with Redis (optional) or in-memory fallback.
 * Install Redis support: npm install redis
 *
 * Usage:
 *     import { cache, cached } from '@/lib/cache';
 *
 *     // Set a value (TTL in seconds)
 *     await cache.set('key:123', value, 300);
 *
 *     // Get a value
 *     const value = await cache.get('key:123');
 *
 *     // Delete a value
 *     await cache.delete('key:123');
 *
 *     // Delete all keys matching pattern (supports * wildcard)
 *     await cache.deletePattern('user:*');        // All user keys
 *     await cache.deletePattern('product:*');     // All product keys
 *
 *     // Use decorator for function caching
 *     @cached(300)
 *     async function getData(dataId: string) {
 *         // This result will be cached for 5 minutes
 *         return await fetchFromDatabase(dataId);
 *     }
 */

import { createClient, RedisClientType } from 'redis';
import crypto from 'crypto';

const REDIS_URL = process.env.REDIS_URL || 'redis://localhost:6379/0';

interface CacheEntry<T> {
  value: T;
  expiresAt: number | null;
}

class InMemoryCache {
  private store: Map<string, CacheEntry<any>> = new Map();

  async get<T>(key: string): Promise<T | null> {
    const entry = this.store.get(key);
    if (!entry) {
      return null;
    }

    if (entry.expiresAt && Date.now() > entry.expiresAt) {
      this.store.delete(key);
      return null;
    }

    return entry.value as T;
  }

  async set<T>(key: string, value: T, ttl?: number): Promise<void> {
    const expiresAt = ttl ? Date.now() + ttl * 1000 : null;
    this.store.set(key, { value, expiresAt });
  }

  async delete(key: string): Promise<void> {
    this.store.delete(key);
  }

  async deletePattern(pattern: string): Promise<void> {
    // Simple wildcard support for in-memory cache
    if (pattern.includes('*')) {
      const [prefix, suffix] = pattern.split('*');
      const keysToDelete: string[] = [];
      
      for (const key of this.store.keys()) {
        if (key.startsWith(prefix) && (!suffix || key.endsWith(suffix))) {
          keysToDelete.push(key);
        }
      }
      
      for (const key of keysToDelete) {
        this.store.delete(key);
      }
    } else {
      // Exact match
      this.store.delete(pattern);
    }
  }

  async exists(key: string): Promise<boolean> {
    const entry = this.store.get(key);
    if (!entry) {
      return false;
    }

    if (entry.expiresAt && Date.now() > entry.expiresAt) {
      this.store.delete(key);
      return false;
    }

    return true;
  }

  async clear(): Promise<void> {
    this.store.clear();
  }

  async close(): Promise<void> {
    // No-op for in-memory
  }
}

class RedisCache {
  private client: RedisClientType | null = null;
  private url: string;

  constructor(url: string) {
    this.url = url;
  }

  private async getClient(): Promise<RedisClientType> {
    if (!this.client) {
      this.client = createClient({ url: this.url });
      await this.client.connect();
    }
    return this.client;
  }

  async get<T>(key: string): Promise<T | null> {
    try {
      const client = await this.getClient();
      const value = await client.get(key);
      
      if (value === null) {
        return null;
      }

      try {
        return JSON.parse(value) as T;
      } catch {
        return value as T;
      }
    } catch (error) {
      console.error('Redis get error:', error);
      return null;
    }
  }

  async set<T>(key: string, value: T, ttl?: number): Promise<void> {
    try {
      const client = await this.getClient();
      const serialized = typeof value === 'string' ? value : JSON.stringify(value);

      if (ttl) {
        await client.setEx(key, ttl, serialized);
      } else {
        await client.set(key, serialized);
      }
    } catch (error) {
      console.error('Redis set error:', error);
    }
  }

  async delete(key: string): Promise<void> {
    try {
      const client = await this.getClient();
      await client.del(key);
    } catch (error) {
      console.error('Redis delete error:', error);
    }
  }

  async deletePattern(pattern: string): Promise<void> {
    try {
      const client = await this.getClient();
      let deletedCount = 0;
      
      for await (const key of client.scanIterator({ MATCH: pattern })) {
        await client.del(key);
        deletedCount++;
      }
      
      if (deletedCount > 0) {
        console.debug(`Deleted ${deletedCount} cache keys matching pattern: ${pattern}`);
      }
    } catch (error) {
      console.error('Redis deletePattern error:', error);
    }
  }

  async exists(key: string): Promise<boolean> {
    try {
      const client = await this.getClient();
      const result = await client.exists(key);
      return result > 0;
    } catch (error) {
      console.error('Redis exists error:', error);
      return false;
    }
  }

  async clear(): Promise<void> {
    try {
      const client = await this.getClient();
      await client.flushDb();
    } catch (error) {
      console.error('Redis clear error:', error);
    }
  }

  async close(): Promise<void> {
    if (this.client) {
      await this.client.quit();
      this.client = null;
    }
  }
}

function createCache(): InMemoryCache | RedisCache {
  // Use Redis if REDIS_URL is set and redis package is available
  if (REDIS_URL && REDIS_URL !== 'redis://localhost:6379/0') {
    try {
      return new RedisCache(REDIS_URL);
    } catch (error) {
      console.warn('Redis not available, using in-memory cache. Install with: npm install redis');
      return new InMemoryCache();
    }
  }

  // Check if redis package is installed
  try {
    require.resolve('redis');
    return new RedisCache(REDIS_URL);
  } catch {
    return new InMemoryCache();
  }
}

export const cache = createCache();

/**
 * Decorator to cache function results.
 *
 * @param ttl Time to live in seconds (default: 5 minutes)
 * @param keyPrefix Prefix for cache key
 */
export function cached<T extends (...args: any[]) => Promise<any>>(
  ttl: number = 300,
  keyPrefix: string = ''
): (target: any, propertyKey: string, descriptor: TypedPropertyDescriptor<T>) => void {
  return function (target: any, propertyKey: string, descriptor: TypedPropertyDescriptor<T>) {
    const originalMethod = descriptor.value!;

    descriptor.value = async function (...args: any[]) {
      // Build cache key
      const argHash = crypto
        .createHash('md5')
        .update(JSON.stringify(args))
        .digest('hex')
        .substring(0, 16);
      
      const prefix = keyPrefix || target.constructor.name || '';
      const cacheKey = `${prefix}:${propertyKey}:${argHash}`;

      // Try to get from cache
      const cachedValue = await cache.get(cacheKey);
      if (cachedValue !== null) {
        return cachedValue;
      }

      // Call function and cache result
      const result = await originalMethod.apply(this, args);
      await cache.set(cacheKey, result, ttl);
      return result;
    } as T;

    return descriptor;
  };
}

/**
 * Helper to invalidate cache by pattern.
 * Supports wildcard patterns like "user:*" or "product:*".
 */
export async function invalidateCache(pattern: string): Promise<void> {
  if (cache.deletePattern) {
    await cache.deletePattern(pattern);
  } else {
    // Fallback for older cache implementations
    await cache.delete(pattern);
  }
}

